# Problem Statement

The goal is to predict the CHURN in the telecom industry. Create the best model which can predict customer swtiching to different telecom. The idea is to retent customer who are leaving by applying the retention stratergy.
This is the classification problem, model predicting whether customer is gonna churned or not.

# Data 

The customer churn data is given in the file data/raw.csv. The detailed explanation is as follows:

- **State**: The state where a customer comes from
- **Account length**: Number of days a customer has been using services
- **Area code**: The area where a customer comes from
- **Phone number**: The phone number of a customer
- **International plan**: The status of customer international plan
- **Voicemail plan**: The status of customer voicemail plan
- **No. vmail msgs**: Number of voicemail message sent by a customer
- **Total day minutes**: Total call minutes spent by a customer during day time
- **Total day calls**: Total number of calls made by a customer during day time
- **Total day charge**: Total amount charged to a customer during day time
- **Total eve minutes**: Total call minutes spent by a customer during evening time
- **Total eve calls**: Total number of calls made by a customer during evening time
- **Total eve charge**: Total amount charged to a customer during evening time
- **Total night minutes**: Total call minutes spent by a customer during night time
- **Total night calls**: Total number of calls made by a customer during night time
- **Total night charge**:Total amount charged to a customer during night time
- **Total intl minutes**: Total international call minutes spent by a customer
- **Total intl calls**: Total number of international calls made by a customer
- **Total int charge**: Total international call amount charged to a customer
- **Customer service calls**: Total number of customer service calls made by a customer
- **Churn**: Whether a customer is churned or not



# Evalution Metric

We will measure the accuracy, precision, f1 Score, recall of the model. we will more focus on getting the recall which is balance between recall and precision.

# import

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
import warnings 
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv("../data/raw.csv")

# Intial Inspection

- check for missing values
- check for dupilcate values
- check outliers in the dataset
- check the dtype for all columns
- check imbalance in the dataset

In [3]:
df.head()

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...  total eve calls  total eve charge  \
0             45.07  ...               99             16.78   
1             27.47  ...              103             16.62   
2             41.38  ...              110             10.30   
3             50.90  ...               88              5.26   
4             28.34  ...              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [4]:
df.sample(10).T

856       2938      1895      226       88    \
state                         WY        NE        HI        MO        KY   
account length               104        31        86       123        74   
area code                    408       415       408       415       415   
phone number            366-3917  338-9044  398-3004  402-6591  346-7302   
international plan            no        no        no        no        no   
voice mail plan               no        no       yes        no       yes   
number vmail messages          0         0        21         0        35   
total day minutes          183.6      97.5     197.9     206.9     154.1   
total day calls              133       129        99        85       104   
total day charge           31.21     16.58     33.64     35.17      26.2   
total eve minutes          120.7     260.4     165.6     244.7     123.4   
total eve calls               98        78       100        78        84   
total eve charge           10.26     22.13     14.08      20.8     10.49   
total night minutes        215.1      88.7     208.0     221.5     202.1   
total night calls            112       100       120       136        57   
total night charge          9.68      3.99      9.36      9.97      9.09   
total intl minutes          12.7       7.0      10.1       7.7      10.9   
total intl calls               2         5         9         2         9   
total intl charge           3.43      1.89      2.73      2.08      2.94   
customer service calls         1         1         0         3         2   
churn                      False     False     False     False     False   

                            962       295       3205      457       1635  
state                         NJ        IN        MS        KY        AZ  
account length               107       166       111        54        66  
area code                    415       510       408       408       510  
phone number            354-9062  350-6759  345-3787  356-1420  337-8618  
international plan            no        no        no        no        no  
voice mail plan               no        no        no        no        no  
number vmail messages          0         0         0         0         0  
total day minutes          212.1     204.2     294.7     159.8     154.0  
total day calls               95       115        90        99       133  
total day charge           36.06     34.71      50.1     27.17     26.18  
total eve minutes          150.1     179.9     294.6     264.0     198.9  
total eve calls               88       152        72        64       121  
total eve charge           12.76     15.29     25.04     22.44     16.91  
total night minutes        219.8     216.8     260.1     115.7     151.9  
total night calls            111       109       121        70       100  
total night charge          9.89      9.76      11.7      5.21      6.84  
total intl minutes           7.7       9.5      10.8       9.7       9.5  
total intl calls               2         5         3         7         3  
total intl charge           2.08      2.57      2.92      2.62      2.57  
customer service calls         3         1         1         2         4  
churn                      False     False      True     False      True

In [5]:
df.shape

(3333, 21)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   state                   3333 non-null   object 
 1   account length          3333 non-null   int64  
 2   area code               3333 non-null   int64  
 3   phone number            3333 non-null   object 
 4   international plan      3333 non-null   object 
 5   voice mail plan         3333 non-null   object 
 6   number vmail messages   3333 non-null   int64  
 7   total day minutes       3333 non-null   float64
 8   total day calls         3333 non-null   int64  
 9   total day charge        3333 non-null   float64
 10  total eve minutes       3333 non-null   float64
 11  total eve calls         3333 non-null   int64  
 12  total eve charge        3333 non-null   float64
 13  total night minutes     3333 non-null   float64
 14  total night calls       3333 non-null   

### Check for nan

In [7]:
df.isna().sum()

state                     0
account length            0
area code                 0
phone number              0
international plan        0
voice mail plan           0
number vmail messages     0
total day minutes         0
total day calls           0
total day charge          0
total eve minutes         0
total eve calls           0
total eve charge          0
total night minutes       0
total night calls         0
total night charge        0
total intl minutes        0
total intl calls          0
total intl charge         0
customer service calls    0
churn                     0
dtype: int64

### check for duplication

In [8]:
print(df.duplicated().sum())

0


### check unique values for columns

In [9]:
df.nunique()

state                       51
account length             212
area code                    3
phone number              3333
international plan           2
voice mail plan              2
number vmail messages       46
total day minutes         1667
total day calls            119
total day charge          1667
total eve minutes         1611
total eve calls            123
total eve charge          1440
total night minutes       1591
total night calls          120
total night charge         933
total intl minutes         162
total intl calls            21
total intl charge          162
customer service calls      10
churn                        2
dtype: int64

### Statistics


In [10]:
df.describe()

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls  
count        3333.000000             3333.000000  
mean            2.764581                1.562856  
std             0.753773                1.315491  
min             0.000000                0.000000  
25%             2.300000                1.000000  
50%             2.780000                1.000000  
75%             3.270000                2.000000  
max             5.400000                9.000000

In [11]:
df.describe(include="object")

state phone number international plan voice mail plan
count   3333         3333               3333            3333
unique    51         3333                  2               2
top       WV     382-4657                 no              no
freq     106            1               3010            2411

### Check for the outlier

In [12]:
def check_outlier_IQR(data):
    '''
    this function check for the no. of outliers for all the numberic column in dataset
    '''
    numberic_column = data.select_dtypes(include='number').columns
    outlier_count = {}

    for col in numberic_column:
        Q1 = data[col].quantile(0.25)
        Q3 = data[col].quantile(0.75)
        IQR = Q3 - Q1
        upper_wisker = Q3 + (1.5 * IQR)
        lower_wisker = Q1 - (1.5 * IQR)
        no_outliers = data[(data[col] > upper_wisker) | (data[col] < lower_wisker)][col].count()

        outlier_count[col] = no_outliers
    
    return outlier_count

In [14]:
outlier_count = check_outlier_IQR(df)
df_outlier_count = pd.DataFrame(list(outlier_count.items()), columns=["columns", "number_of_outliers"])
df_outlier_count

columns  number_of_outliers
0           account length                  18
1                area code                   0
2    number vmail messages                   1
3        total day minutes                  25
4          total day calls                  23
5         total day charge                  25
6        total eve minutes                  24
7          total eve calls                  20
8         total eve charge                  24
9      total night minutes                  30
10       total night calls                  22
11      total night charge                  30
12      total intl minutes                  46
13        total intl calls                  78
14       total intl charge                  49
15  customer service calls                 267

### Check for the Imbalance 


In [15]:
churn_value_count = df["churn"].value_counts().reset_index()
churn_value_count["Weightage"] = churn_value_count["count"] / df.shape[0]
churn_value_count

churn  count  Weightage
0  False   2850   0.855086
1   True    483   0.144914

### Conlusion - initial inpection

From the initial inpection, we can conlude that:

- There are 3333 rows and total 21 columns in the dataset
- Out of 3333, 3010 do not have International plan: 90.30%
- Out of 3333, 2411 do not have Voice mail plan: 72.33%
- The dataset is highly imbalance, 85.5% customer didn't churn whearas 14.5% did. 
- A lot of columns has ouliter, we need to deal with that, as outlier baised the result and perform poorly in genrealization and overtfit the model


# Data Preprocessing 

- deal with missing Values and duplicates - there aren't any missing values niether dulpicates
- Data type correction
- Drop phone number (doen't add any values in predicting chrun)

In [16]:
# Data type correction
df["area code"] = df["area code"].astype("object")
df.describe(include="object")

state  area code phone number international plan voice mail plan
count   3333       3333         3333               3333            3333
unique    51          3         3333                  2               2
top       WV        415     382-4657                 no              no
freq     106       1655            1               3010            2411

- Area code is now and categorical column
- Around 50 percent customer is from '415' area code 

In [17]:
# drop phone number
df.drop(columns=["phone number"], inplace=True)
df.shape

(3333, 20)

# EDA 

- distribution of Categorical values 
- distribution of numberic columns
- correlation plots
- pairplot 
- Explore features with respect to target